crypto/asn1/tasn_enc.c

```diff
@@ -173,17 +177,18 @@ int ASN1_item_ex_i2d(ASN1_VALUE **pval, unsigned char **out,
         ASN1_put_object(out, ndef, seqcontlen, tag, aclass);
         for (i = 0, tt = it->templates; i < it->tcount; tt++, i++) {
             const ASN1_TEMPLATE *seqtt;
-            ASN1_VALUE **pseqval;
+            const ASN1_VALUE **pseqval;
             seqtt = asn1_do_adb(pval, tt, 1);
             if (!seqtt)
                 return 0;
-            pseqval = asn1_get_field_ptr(pval, seqtt);
+            pseqval = asn1_get_const_field_ptr(pval, seqtt);
             /* FIXME: check for errors in enhanced version */
             asn1_template_ex_i2d(pseqval, out, seqtt, -1, aclass);
         }
         if (ndef == 2)
             ASN1_put_eoc(out);
-        if (asn1_cb && !asn1_cb(ASN1_OP_I2D_POST, pval, it, NULL))
+        /* ASN1_OP_I2D_POST appears unused */
+        if (asn1_cb && !asn1_cb(ASN1_OP_I2D_POST, (ASN1_VALUE **)pval, it, NULL))
```

Comment:
This cast looks potentially dangerous....couldn't this cb come from user code which could do anything with it? (Same comment for a number of other similar places)

Reply:
I'm aware of that, but AFAICS so far this code has no effect anyway because `ASN1_OP_I2D_POST` is not really used by any ASN.1 type, which I tried to make explicit with the new comment.

A clean solution would be either to remove the `ASN1_OP_I2D_POST` option for the `ASN1_aux_cb *asn1_cb` or to use a separate callback type for the `ASN1_OP_I2D_POST` operation. 

I also had investigated the possibility to constify the second parameter (`ASN1_VALUE **in`) in
```
typedef int ASN1_aux_cb(int operation, ASN1_VALUE **in, const ASN1_ITEM *it,
                        void *exarg);
```
but it looks like this this would require constification of the third parameter of the following function in `crypto/asn1/asn_mime.c`, which unfortunately is not possible because the way `CMS_dataFinal()` is coded prevents this. Therefore I had added a respective comment:
```
/* cannot constify val because of CMS_dataFinal() */
static int asn1_output_data(BIO *out, BIO *data, ASN1_VALUE *val, int flags,
                            const ASN1_ITEM *it)
```
I also had added similar comments regarding `CMS_stream()`.

Apparently larger portions of the CMS code would need some deeper cleanup.

Reply:
It's not used by any built-in type but it is possible that applications that define their own ASN.1 types could use it. Removing it or introducing a separate callback type would represent an API change - which we shouldn't be doing for 3.0 (i.e. master)

Reply:
Meanwhile I've found that the third `ASN1_aux_cb()` parameter can - and should - be `const` for `ASN1_OP_I2D_PRE`, `ASN1_OP_I2D_POST`, `ASN1_OP_PRINT_PRE`, and `ASN1_OP_PRINT_POST` (and likely not for and other of the defined operations).

Since none of these four operations are used so far within OpenSSL it would be easy to define
```
typedef int ASN1_aux_const_cb(int operation, const ASN1_VALUE **in,
                              const ASN1_ITEM *it, void *exarg);
```
a add to `struct ASN1_AUX_st` the following member:
```
    ASN1_aux_const_cb *asn1_const_cb;
```
and initialize it by default to `NULL`.
I've tentatively done this in this PR and tested it.
This would remove the need for the type cast(s) discussed here.

Yes, this implies an API change, but just a minor one: the addition of a new field to a rather internal data structure. Would this really be a problem, in particular since the whole ASN1_AUX feature is not documented?

Reply:
I see two options here:

1.  introduce a separate callback type as detailed above and implemented tentatively in the current state of this PR
2. Use a couple of casts to `const` in `crypto/asn1/tasn_{enc,prn}.c`

What do you prefer?

Reply:
@mattcaswell commented on Feb 7:
> This represents an API break since anyone with old code that only defines `asn1_cb` (and not `asn1_const_cb`) will suddenly stop working - won't it??

Potentially yes, but note that, as mentioned before, this callback is used only for `ASN1_OP_I2D_PRE`, `ASN1_OP_I2D_POST`, `ASN1_OP_PRINT_PRE`, and `ASN1_OP_PRINT_POST` and none of these four operations are used so far within OpenSSL. Moreover the whole ASN_AUX_feature so far is undocumented. ThusI doubt it is used in any critical way so far, and tend to think that this API change is a minor, bearable one.

Again, the alternative is to introduce a couple of casts to `const` (as in the original version of this PR), and the question I put on Feb 23 is still open, namely which of the two option is preferred.

Reply:
Please include links to comments you refer to.  Github says "xx days ago" for quite a lot of them.  The time stamp of each comment is a permalink, please use that.

We really do need to be careful not to be overly optimistic about what's not used, you can't really know, and for the lack of documentation, some people do study the header files quite closely.

The ASN1_AUX structure has a `flags` field, please consider adding a flag that can be used by new code to signal that newer fields exist, and adapt the code to handle cases where that flag isn't set.  This is a technique we used for exactly this sort of thing in the RSA method, see here (from the 1.0.2 branch):

https://github.com/openssl/openssl/blob/fff469b269d8309377291ff86767314d7489fd84/crypto/rsa/rsa.h#L108-L115

Reply:
> Please include links to comments you refer to. Github says "xx days ago" for quite a lot of them. The time stamp of each comment is a permalink, please use that.

All right, I'll start using the permalinks. 
Actually I find Github's "xx days ago" style a bit inconvenient, too. 
Luckily, when hovering the mouse pointer over such references, the exact (absolute) time is shown.

Reply:
> We really do need to be careful not to be overly optimistic about what's not used, you can't really know, and for the lack of documentation, some people do study the header files quite closely.

Good points.

> The ASN1_AUX structure has a `flags` field, please consider adding a flag that can be used by new code to signal that newer fields exist, and adapt the code to handle cases where that flag isn't set.

I've done this now, adding a `ASN1_AFLG_CONST_CB` flag, which indicates that for the `ASN1_OP_I2D_PRE`, `ASN1_OP_I2D_POST`, `ASN1_OP_PRINT_PRE`, and `ASN1_OP_PRINT_POST` operations a callback function with `const ASN1_VALUE **in` parameter is present.